utils/result_loop.hh

```diff
@@ -251,4 +252,52 @@ result_map_reduce(Iterator begin, Iterator end, Mapper&& mapper, Initial initial
             internal::result_map_reduce_binary_adapter<left_type, right_type, Reducer>(std::move(reduce)));
 }
 
+template<typename AsyncAction, typename StopCondition>
+requires requires (StopCondition cond, AsyncAction act) {
+    { cond() } -> std::same_as<bool>;
+    { seastar::futurize_invoke(act) } -> ExceptionContainerResultFuture<>;
+}
+inline
+auto
+result_do_until(StopCondition stop_cond, AsyncAction action) {
+    // TODO: Constrain the result of act() better
+    using future_type = seastar::futurize_t<std::invoke_result_t<AsyncAction>>;
+    using result_type = typename future_type::value_type;
+    static_assert(std::is_void_v<typename result_type::value_type>,
+            "The result type of the action must be future<result<void>>");
+    for (;;) {
```

Comment:
do_until() is obsoleted by coroutines (it's a while loop). How would a coroutine using results look like? Isn't it just a while loop with an extra check on the result?

The only advantage do_until() has over a coroutine is zero allocations  if we never defer, but we give that up by using a coroutine here. I suggest to drop this and open-code it in the caller. It risks the caller not checking for errors, but that's a common risk with result types.

Reply:
Sorry, I see you carefully avoided the allocation in case everything was ready.

So it really depends on the callers. If they are eligible for coroutine conversion (using do_with or expected to block), then they should be using coroutines. Otherwise, this.